src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Kafka/KafkaProduceSyncDeliveryHandlerIntegration.cs

```diff
@@ -0,0 +1,182 @@
+﻿using System;
+using System.Reflection;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Kafka
+{
+    /// <summary>
+    /// Confluent.Kafka Producer.TypedDeliveryHandlerShim_Action.HandleDeliveryReport calltarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "Confluent.Kafka",
+        TypeName = "Confluent.Kafka.Producer`2+TypedDeliveryHandlerShim_Action",
+        MethodName = ".ctor",
+        ReturnTypeName = ClrNames.Void,
+        ParameterTypeNames = new[] { ClrNames.String, "!0", "!1", KafkaConstants.ActionOfDeliveryReportTypeName },
+        MinimumVersion = "1.4.0",
+        MaximumVersion = "1.*.*",
+        IntegrationName = KafkaConstants.IntegrationName)]
+    public class KafkaProduceSyncDeliveryHandlerIntegration
+    {
+        private static readonly IDatadogLogger Logger = DatadogLogging.GetLoggerFor<KafkaProduceSyncDeliveryHandlerIntegration>();
+
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <typeparam name="TKey">Type of the message key</typeparam>
+        /// <typeparam name="TValue">Type of the message value</typeparam>
+        /// <typeparam name="TActionOfDeliveryReport">Type of the delivery report</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="topic">The topic to which the message was sent</param>
+        /// <param name="key">The message key value</param>
+        /// <param name="value">The message value</param>
+        /// <param name="handler">The delivery handler instance</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TTarget, TKey, TValue, TActionOfDeliveryReport>(TTarget instance, string topic, TKey key, TValue value, TActionOfDeliveryReport handler)
+        {
+            if (handler is null)
+            {
+                // Handled in KafkaProduceSyncIntegration
+                return CallTargetState.GetDefault();
+            }
+
+            try
+            {
+                var span = Tracer.Instance?.ActiveScope?.Span;
+                if (span is null)
+                {
+                    Logger.Warning("Unexpected null span for Kafka Producer with delivery handler");
+                    return CallTargetState.GetDefault();
+                }
+
+                var newAction = CachedWrapperDelegate<TActionOfDeliveryReport>.CreateWrapper(handler, span);
+
+                Action<ITypedDeliveryHandlerShimAction> updateHandlerAction = inst =>
+                {
+                    try
+                    {
+                        inst.Handler = newAction;
```

Comment:
I don't understand why we need to re-set the Handler field inside of a new Action instead of just executing the code. Is there a specific reason for that? If we can avoid that, then that's one fewer allocation

Reply:
I _think_ we have to do it this way. 

We need to set the handler in `MethodEnd`, not in `MethodBegin`. But we have to generate the `newAction` variable in `MethodBegin` as that's where we have access to the `TActionOfDeliveryReport` type. 

With the approach you're suggesting, we would need to pass the `newAction` as state in `CallTargetState`. But then we wouldn't be able to `cast` it to the correct type in `MethodEnd`, as we don't have the `TActionOfDeliveryReport` generic parameter there. 

So I _think_ this approach is the only one that is really doable - I couldn't find a way to set it directly in `MethodEnd` without either losing the benefits of `CachedWrapperDelegate`, or having to set the `Handler` field through reflection?

Reply:
You're right, given all the constraints this does seem like the best way to do it. However, what if instead of instrumenting the constructor of `Confluent.Kafka.Producer`2+TypedDeliveryHandlerShim_Action` we instead just instrument the `HandleDeliveryReport` method itself? Looking at the decompilation for `Confluent.Kafka 1.4.3.0`, all the fields we are reading come straight from the `TopicPartitionOffsetError` property of the `DeliveryReport` argument and these values look to be the same when they're eventually passed to the handler. That could remove a lot of this complexity and possibly some allocations. Have you already tried this approach and, if so, were there some gotcha's there?

Reply:
Absolutely agree this would be an easier approach, and we did consider it. The main problems with this approach are:

1. You lose the correct "start time" for the produce span
2. More critically, we'd lose the correct _parent_ span, given that the callback runs once the `Produce()` method has finished, and potentially a different span is active..
3. We don't have access to the message Value in the callback, so wouldn't be able to tag Tombstones any more

No. 2 is the big problem, and the main reason we went for the current, more complex, approach. Does that make sense to you?

Reply:
Got it, I didn't realize how crazy this instrumentation scenario would be. Thanks for taking the time to respond to everything 😄